File: src/main/java/com/ml/atomic/mathgame/CalcStep.java
package com.ml.atomic.mathgame;

import com.google.common.collect.Range;
import lombok.AllArgsConstructor;
import lombok.ToString;
import lombok.extern.slf4j.Slf4j;

import java.util.Optional;

import static org.apache.commons.lang3.RandomUtils.nextInt;

@Slf4j
@AllArgsConstructor @ToString
public final class CalcStep {
    public static final int MAX_ERRORS = 999;

    Operator operator;

    /**
     * 被操作数的闭区间
     */
    Range<Integer> numRange;

    /**
     * 分步计算结果闭区间
     */
    Range<Integer> resultRange;

    /**
     * 单步计算，计算到符合限制的数据为止
     */
    public Optional<Result> apply(int lastNumber) {
        int errorCount = 0;
        do {
            int nextNumber = nextInt(numRange.lowerEndpoint(), numRange.upperEndpoint());
            int result;
            switch (this.operator) {
                case ADD: {
                    result = lastNumber + nextNumber;
                    break;
                }
                case SUB: {
                    result = lastNumber - nextNumber;
                    break;
                }
                default:
                    throw new IllegalArgumentException("使用了暂未支持的操作符:" + this.operator.label);
            }

            if (this.resultRange.contains(result)) return Optional.of(new Result(nextNumber, result, this.operator));
            else errorCount++;

            // 超过最大重试次数
            if (errorCount >= MAX_ERRORS) {
                System.out.println(String.format("calc false... %s %s ? in %s", lastNumber, this.operator.label, resultRange));
                return Optional.empty();
            }

        } while (true);
    }

    @AllArgsConstructor
    static class Result {

        /**
         * 分步参算值
         */
        int val;

        /**
         * 单步计算结果(最后一步的结果可以直接输出，其他的可以用于打印)
         */
        int result;

        /**
         * 操作符
         */
        Operator operator;
    }

}

File: calc_step.py
from typing import Optional, TypedDict
from random import randint

import portion

from opt import Operator


class Result(TypedDict):
    """
    计算结果的holder
    """
    operator: Operator
    val: int
    result: int


class CalcStep(object):
    """
    计算步骤的抽象
    """

    # 当循环求值重试次数> max_error时，中止本轮计算
    max_error: int = 999

    def __init__(self, operator: Operator, num_range: portion, result_range: portion):
        """
        :param num_range 参与运算的数的区间
        :param result_range 结果的区间
        """

        self.operator = operator
        self.num_range = num_range
        self.result_range = result_range

    def apply(self, last_number: int) -> Optional[Result]:
        """
        :param last_number 上一个数
        """
        error_count = 0
        while True:
            next_number = randint(self.num_range.lower, self.num_range.upper)
            result = -1
            if self.operator == Operator.ADD:
                result = last_number + next_number
            elif self.operator == Operator.SUB:
                result = last_number - next_number
            else:
                raise ValueError("使用了暂未支持的操作符:", self.operator)

            if result in self.result_range:
                return {"operator": self.operator, "val": next_number, "result": result}
            else:
                error_count += 1

            if error_count >= CalcStep.max_error:
                print("求值失败", last_number, " ", self.operator.value,
                      " ? in", self.result_range)
                return None
